useBlockPreview: Try outputting EditorStyles to ensure local style overrides are rendered

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes: #54956

Explores an alternative to #55241 that instead of syncing style overrides, tries to output `EditorStyles` instead. This PR fixes an issue where Duotone styles set within a Post Template block are not rendered in non-selected instances of a Query Loop within the post or site editors. It also fixes an issue where Duotone styles in Safari were not updating for these previews, either.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Prior to this PR, for things that use `useBlockPreview`, the styles that are generated within that preview were not output anywhere. This is because `ExperimentalBlockEditorProvider` uses its own registry, so is designed to have values passed down to it, but not to propagate values outside of itself. So any calls to `setStyleOverride` would update state within the local block editor provider, and not be captured by the parent's `EditorStyles`.

The fix in this PR is to output an `EditorStyles` component within the preview, and ensure that the local-to-the-preview block editor state only contains local styles, so we avoid and duplicates between the preview and the parent block editor environment.

There is a small drawback that the `style` and `svg` tags in this PR are rendered as a direct child of the post template block's `li` element, however based on feedback on this PR and #55241, it seems that this drawback is not as bad as the difficulty of accurately syncing between `overridesStyles` states, which was explored in #55241.

This PR is currently my preferred approach to try to fix this bug.

<!--
Drawbacks of this approach:

* We now have additional `style` and `svg` tags being output within the same document, and they're inserted as a direct child of the `li` element of the post template block, which could conflict with theme styles:
<img width="967" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/d01eb014-692b-45dc-80c5-3265ea110bed">

* There are now duplicate `filter` ids being output because of duplicate `svg` elements being output:

<img width="399" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/cab8cbe8-e186-42a4-873d-83cb3cbe9209">

* Safari behaviour appears to be degraded as styles do not always update (though #55241 also exhibited bugs). In Safari, the first update of the Duotone styles appear to work, but subsequent ones don't (this is possibly related to duplicate `id`s, too):

https://github.com/WordPress/gutenberg/assets/14988353/fdcf0c05-1428-454c-8198-8f912f3012b7

Update: upon further testing, it appears this buggy behaviour in Safari is present in both this fix and over in #55241 🤔

-->

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Try using `EditorStyles` to output styles for the `useBlockPreview`'s local styles instead of syncing as in #55241.
* In the `useBlockPreview`'s settings that get passed down to `ExperimentalBlockEditorProvider`, strip it of `styles` so that any styles applied in the parent environment are not duplicated in the child preview's editor styles. This ensures that only styles generated within that preview are output via `EditorStyles`.
* In the Duotone hook, add some logic to repaint the block that uses Duotone styles, if the current session is running Safari. While we usually like to avoid browser sniffing, this ensures that the slight performance hit is only encountered in browsers where the repaint hack is needed.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Add a couple of posts with a featured image set.
2. In another post, page, or template, add a Query Loop / Post Template block containing featured images and at least 2 items per page (you can use the test markup below).
3. In the editor, prior to this PR, only the selected item within the Post Template will render Duotone styles.
4. With this PR applied, each item in the Post Template loop should render Duotone styles.
5. If you go to adjust the Duotone styles for any of the blocks in the loop, it should update in real-time (with a slight delay as styles are propagated).
6. Test adding a bunch of other images to the same post outside of the loop, and use the Duotone filter, with custom values for the background and foreground colours. There should be no clash with the styles updated from with the Post Template loop 🤞
7. Test in the site editor as well as the post editor, and a range of browsers (including Safari), and in different block configurations (e.g. Image block as a child of Row, etc)


<details>

<summary>Test markup</summary>

```html
<!-- wp:query {"queryId":9,"query":{"perPage":"6","pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false}} -->
<div class="wp-block-query"><!-- wp:post-template {"layout":{"type":"default","columnCount":3}} -->
<!-- wp:group {"style":{"elements":{"heading":{"color":{"text":"#f44343"}},"link":{"color":{"text":"var:preset|color|primary"}}},"color":{"text":"#4f0808"},"border":{"color":"#929df3","width":"3px","radius":"10px"}},"layout":{"type":"constrained"}} -->
<div class="wp-block-group has-border-color has-text-color has-link-color" style="border-color:#929df3;border-width:3px;border-radius:10px;color:#4f0808"><!-- wp:columns {"verticalAlignment":"center"} -->
<div class="wp-block-columns are-vertically-aligned-center"><!-- wp:column {"verticalAlignment":"center","width":"25%"} -->
<div class="wp-block-column is-vertically-aligned-center" style="flex-basis:25%"><!-- wp:post-featured-image {"isLink":true,"aspectRatio":"auto","style":{"color":{"duotone":["rgb(119, 59, 139)","#fcff41"]}}} /--></div>
<!-- /wp:column -->

<!-- wp:column {"verticalAlignment":"center","width":"75%"} -->
<div class="wp-block-column is-vertically-aligned-center" style="flex-basis:75%"><!-- wp:heading {"level":6} -->
<h6 class="wp-block-heading">A heading</h6>
<!-- /wp:heading -->

<!-- wp:post-title {"isLink":true} /--></div>
<!-- /wp:column --></div>
<!-- /wp:columns --></div>
<!-- /wp:group -->
<!-- /wp:post-template --></div>
<!-- /wp:query -->
```

</details>

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

### Before

Duotone is only applied to the selected item (as the others use a preview whose styles are not being propagated)

https://github.com/WordPress/gutenberg/assets/14988353/5930ecd6-0136-4cca-b0bf-44600aef6f64

### After

Duotone is applied to all items, and the non selected items are updated to reflect the change. Adjusting Duotone for an image outside of the Post Template loop is unaffected:

https://github.com/WordPress/gutenberg/assets/14988353/806c259b-9816-4b10-8ca5-9352aba4e9f5

andrewserong committed:
useBlockPreview: Try alternative fix for displaying local style overrides

ellatrix commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -92,6 +93,17 @@ export function BlockPreview( {
+		};
+	}, [] );
+
+	return <EditorStyles styles={ styles } />;
```
How are these styles being added right now, or previously before the bug?
Why does EditorStyles not include these already?

andrewserong replied:
Prior to the bug being introduced the styles were output via a portal, so even in a child preview with its own block editor provider, the generated styles were being output by the parent. However, since moving over to the block editor store-based state approach for style overrides, this preview no-longer propagates (or renders) any styles generated by the blocks within its preview, because the local-to-the-preview state isn't being used anywhere. Or to put it differently `setStyleOverride` will only update the preview's block editor store, and not the parent environment's block editor store, which is where `EditorStyles` is currently used.

The proposed fix in this PR is to ensure that the preview also outputs `EditorStyles`, but only those styles generated from within the preview.

I've updated the PR description with the following:

Prior to this PR, for things that use useBlockPreview, the styles that are generated within that preview were not output anywhere. This is because ExperimentalBlockEditorProvider uses its own registry, so is designed to have values passed down to it, but not to propagate values outside of itself. So any calls to setStyleOverride would update state within the local block editor provider, and not be captured by the parent's EditorStyles.

andrewserong committed:
Avoid duplicate styles, fix rendering issues in Safari

andrewserong commented:
Update: I think I've gotten this PR working pretty well — here's a screengrab from testing it using Safari:

https://github.com/WordPress/gutenberg/assets/14988353/806c259b-9816-4b10-8ca5-9352aba4e9f5

I've done the following:

* Override `styles` being passed down to `useBlockPreview` so that it's `undefined` — this ensures that the child `useBlockPreview` only renders styles generated from within its own block editor provider's store.
* Added fixes for Safari to force a repaint whenever the Duotone values change, thanks @ajlende for the tip there. I've copied the approach from `duotone.php` with a tiny tweak to not set display to `none` as it affected scroll position: https://github.com/WordPress/gutenberg/blob/9189c3a34714352b92d19a3ef33149b76d8efe64/lib/block-supports/duotone.php#L524-L539

Overall, I think this PR improves the behaviour quite a bit, but it is still a fairly hacky approach so will likely need to be thoroughly tested manually to ensure it doesn't break anything. I've updated the PR description and testing instructions, but please let me know if there's anything I've missed 🙂

tellthemachines commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -113,7 +125,11 @@ export function useBlockPreview( { blocks, props = {}, layout } ) {
-		() => ( { ...originalSettings, __unstableIsPreviewMode: true } ),
+		() => ( {
+			...originalSettings,
+			styles: undefined,
```
To make sure I understand what's happening here, are the styles being set to undefined in this settings object the parent styles? Or the default children styles?

andrewserong replied:
This is clearing out the styles from the parent, so that the child starts off with a blank slate when it comes to the styles within the preview. I'll add an inline comment for this line so that it's clearer what's happening here.

andrewserong replied:
I've added a couple of comments in https://github.com/WordPress/gutenberg/pull/55288/commits/f2ca82166d3e343403ab9a7436005801a6c238f9 to hopefully clarify this a little better 🙂

tellthemachines replied:
Nice, thanks!

tellthemachines reviewed (commented):
Thanks for persisting with this! The fix is working well for me across all platforms and the code change looks very clean.

My only question is if the Safari duotone fix should strictly be a part of this PR and the backport, given that it's an unrelated bug that was already around in the previous release.

andrewserong commented:
> My only question is if the Safari duotone fix should strictly be a part of this PR and the backport, given that it's an unrelated bug that was already around in the previous release.

Good question! From re-testing 6.3.1 without the Gutenberg plugin active, it looks like the previews were working in Safari, so it seems we need to include the Safari fixes in order for 6.4 to be consistent with the working behaviour in 6.3, as far as I can tell. But let me know if you've encountered that bug in 6.3.

Here's a quick test video of 6.3 with Gutenberg deactivated:

https://github.com/WordPress/gutenberg/assets/14988353/bfe060fb-e624-4a30-9412-0dbf7e1c2777

One thing that is slightly unfortunate about the current state of the APIs is that it seems 6.3 was likely a little more performant than this fix. But short of reverting the `setStyleOverride` API, I'm not sure if we have a better option than this PR for 6.4 at this stage 🤔

andrewserong committed:
Add more explanatory comments

ramonjd commented:
For what it's worth, tested with:

- Firefox
- Chrome
- Edge
- Safari
- IOS Safari

And it's working as described.

tellthemachines commented:
> But let me know if you've encountered that bug in 6.3.

It's hard to tell if it's exactly the same bug because the whole preview behaviour is so badly broken on trunk 😬  Instead I was testing switching between duotone filters in Image blocks. On both trunk and 6.3.2, if you apply a filter and then switch to different colors two or three times, it freezes on the first color (on save, the front end updates as expected). That behaviour is fixed on this branch, so I assume that what that Safari fix addressed is an existing bug. 

Having said that, it seems to be a pretty safe change, so not a dealbreaker either way! I thought I'd mention it because we're getting pretty close to RC1 so it's best to be conservative about changes.

andrewserong commented:
> That behaviour is fixed on this branch, so I assume that what that Safari fix addressed is an existing bug.

Ah, gotcha 👍
Yes, it makes sense that if there's any other existing Safari issues due to stale duotone styles not being updated the fix here will probably address it, which means this PR could be stretching slightly beyond the scope of just the bug raised for 6.4.

> Having said that, it seems to be a pretty safe change, so not a dealbreaker either way! I thought I'd mention it because we're getting pretty close to RC1 so it's best to be conservative about changes.

Thank you for mentioning it! Good to be cautious this close to a release and make sure whatever we land is what we're all comfortable with. Let's see what kind of feedback we get for this PR over the weekend 🙂

youknowriad commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -92,6 +93,20 @@ export function BlockPreview( {
 export default memo( BlockPreview );
 
+function InnerStyles() {
+	const { styles } = useSelect( ( select ) => {
```
Isn't `styles` empty here since we're passing an empty styles config to the provider?

andrewserong replied:
~Nope! These `styles` are the styles that are local to the preview, so they're the "real" ones that we wish to output. They've been cleared out of the parent styles, so now have only those styles generated within this local-to-the-preview block editor store.~

Edit: no, you're right! Apologies I totally misunderstood how `styles` was working there. I'll remove it, as I see now `EditorStyles` handles the style overrides internally 🤦

Thanks for the tip! 🙇

andrewserong committed:
Remove additional check for styles within the block preview, as it is not needed since EditorStyles handles its own style overrides retrieval

andrewserong commented:
I've done one more tiny update based on feedback from Riad in https://github.com/WordPress/gutenberg/pull/55288/commits/f07605589cb301bfbf0bd3320c15efc75650f732

I think this should be ready for a final review now 🤞

ramonjd reviewed (approved):
Retested in major browsers list, inc Safari IOS.

This is working well, and there's no perceptable performance hit in Safari.

Also, multiple query blocks on a page preserve their individual block settings:

<img width="235" alt="Screenshot 2023-10-16 at 10 18 23 am" src="https://github.com/WordPress/gutenberg/assets/6458278/284582c9-25a3-4e16-887b-5aefed163857">

andrewserong commented:
Thanks for the re-review @ramonjd and for all the feedback across both explorations, everyone!

Since this fix has wound up being fairly small code-wise, and that it's testing well, and the current approach is consistent with all the feedback so far, I think this is in a good place to land.

I'll merge it in now, but it should be a straightforward revert if we encounter any issues.

Thanks again! 🙇

SiobhyB commented:
I just cherry-picked this PR to the 6.4-rc1-2 branch to get it included in the next release: https://github.com/WordPress/gutenberg/pull/55348/commits/fe74c656741e0daf00c35a35bcd4ef7237e46379

ellatrix commented on packages/block-editor/src/hooks/duotone.js:
```diff
@@ -294,12 +308,30 @@ function DuotoneStyles( {
+			// eslint-disable-next-line no-unused-expressions
+			blockElement.offsetHeight;
+			blockElement.style.display = display;
+		}
```
This is very annoying. How is this related to the original issue? Has this always been an issue in safari? Would be great to find a fix that doesn't rely on blockElement.

ajlende replied:
Safari has a history of not correctly updating HTML elements with SVG filters applied. The approach used here of accessing `el.offsetHeight` was previously used when Safari wouldn't update the element correctly even on the first pass of rendering the page. That particular bug has been fixed now in Safari, but I guess it's broken now for updating CSS after the first pass.

It would be worthwhile to report this issue to https://bugs.webkit.org and add it to the list in https://github.com/WordPress/gutenberg/issues/47302.

andrewserong replied:
Just checked the list of open webkit issues, and it looks like it's already tracked (from way back in 2012!) in this one: https://bugs.webkit.org/show_bug.cgi?id=99996 so I've added it to the list in #47302 

> This is very annoying. How is this related to the original issue? Has this always been an issue in safari? Would be great to find a fix that doesn't rely on blockElement.

It is annoying that we have to include browser-specific workarounds like this, but for now I think keeping this within Duotone and only when the styles are updated is probably the right place for it (at least for WP 6.4). As far as I can tell it's always been an issue in Safari, but the changes from outputting `style` tags via a portal to using the state-based styles in WP 6.4 appears to have made the issue more prominent. At least we've got the workaround in place for now... thanks again for the troubleshooting tips @ajlende!
